node/src/components/consensus/highway_core/highway.rs

```diff
@@ -574,6 +572,19 @@ impl<C: Context> Highway<C> {
         evidence_effects.extend(self.on_new_unit(&unit_hash, now, rng));
         evidence_effects
     }
+
+    /// Adds endorsements to the state.
+    fn add_endorsements(&mut self, endorsements: Endorsements<C>) -> Vec<Effect<C>> {
+        // TODO: Not faulty — evidence!
+        let old_faulty: HashSet<_> = self.state.faulty_validators().collect();
+        self.state.add_endorsements(endorsements, &self.instance_id);
```

Comment:
Ayay, I don't like how `add_endorsements` secretly adds evidences :grimacing: Can you remodel it so that it is more obvious what's happening?

Reply:
Could we, for example, find the conflicting endorsements first, add evidences to the local state, filter them out from the rest of the endorsements and then add the rest?

Reply:
Maybe… but it would be a very nontrivial change, too. I dont' think I'll be able to finish it this week, but maybe we should ticket it.

Reply:
Why? We have access to `state` here, we should be able to call `find_conflicting_endorsements` and return them here.

Reply:
Good point! I'll give it a try tomorrow.